core/src/main/java/org/elasticsearch/cockroach/CockroachActionExecutor.java

```diff
@@ -0,0 +1,363 @@
+                // something went wrong - unregistering task
+                taskManager.unregister(task);
+            }
+        }
+    }
+
+    private void finishTask(String uuid) {
+        RunningCockroachTask task = runningTasks.remove(uuid);
+        if (task != null && task.getTask() != null) {
+            taskManager.unregister(task.getTask());
+        }
+    }
+
+
+    private <Response extends CockroachResponse> void restartNotification(RunningCockroachTask<Response> task) {
+        logger.trace("resending notification for task {}", task.getUuid());
+        if (task.restartNotification()) {
+            if (task.getResponse() != null) {
+                cockroachTransportService.sendResponse(
+                    task.getTask(), task.getUuid(), task.getResponse(), new PublishingFailureListener<>(task));
+            } else {
+                cockroachTransportService.sendFailure(
+                    task.getTask(), task.getUuid(), task.getFailure(), new PublishingFailureListener<>(task));
+            }
+        } else {
+            logger.warn("attempt to resend notification for task {} in the {} state", task.getUuid(), task.getState());
+        }
+    }
+
+    private <Response extends CockroachResponse> void startNotification(RunningCockroachTask<Response> task, Exception e) {
+        logger.trace("sending notification for failed task {}", task.getUuid());
+        if (task.startNotification(e)) {
+            cockroachTransportService.sendFailure(task.getTask(), task.getUuid(), e, new PublishingFailureListener<>(task));
+        } else {
+            logger.warn("attempt to send notification for task {} in the {} state", task.getUuid(), task.getState());
+        }
+    }
+
+    private <Response extends CockroachResponse> void startNotification(RunningCockroachTask<Response> task, Response response) {
+        logger.trace("sending notification for finished task {}", task.getUuid());
+        if (task.startNotification(response)) {
+            cockroachTransportService.sendResponse(task.getTask(), task.getUuid(), response, new PublishingFailureListener<>(task));
+        } else {
+            logger.warn("attempt to send notification for task {} in the {} state", task.getUuid(), task.getState());
+        }
+
+    }
+
+    private class CockroachTaskListener<Response extends CockroachResponse> implements ActionListener<Response> {
+        private final RunningCockroachTask<Response> task;
+
+        public CockroachTaskListener(final RunningCockroachTask<Response> task) {
+            this.task = task;
+        }
+
+        @Override
+        public void onResponse(Response response) {
+            startNotification(task, response);
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            startNotification(task, e);
+        }
+    }
+
+    private class PublishingFailureListener<Response extends CockroachResponse> implements ActionListener<Empty> {
```

Comment:
perhaps a different name for this listener, as it doesn't only handle failures but also successful response publishing